lib/timeline-window.js

```diff
@@ -0,0 +1,392 @@
+        debuglog("TimelineWindow: increased cap by " + count +
+                 " (now " + this._eventCount + ")");
+        // remove some events from the other end, if necessary
+        if (this._eventCount > this._windowLimit) {
+            this._unpaginate(direction != EventTimeline.BACKWARDS);
+        }
+        return q(true);
+    }
+
+    if (!makeRequest) {
+        return q(false);
+    }
+
+    // try making a pagination request
+    var token = tl.timeline.getPaginationToken(direction);
+    if (!token) {
+        debuglog("TimelineWindow: no token");
+        return q(false);
+    }
+
+    debuglog("TimelineWindow: starting request");
+    var self = this;
+    var prom = this._client.paginateEventTimeline(tl.timeline, {
+        backwards: direction == EventTimeline.BACKWARDS,
+        limit: size
+    }).finally(function() {
+        tl.pendingPaginate = null;
+    }).then(function(r) {
+        debuglog("TimelineWindow: request completed with result " + r);
+        if (!r) {
+            // end of timeline
+            return false;
+        }
+        return self.paginate(direction, size);
+    });
+    tl.pendingPaginate = prom;
+    return prom;
+};
+
+
+/**
+ * Trim the window to the windowlimit
+ *
+ * @param {boolean} startOfTimeline if events should be removed from the start
+ *     of the timeline.
+ *
+ * @private
+ */
+TimelineWindow.prototype._unpaginate = function(startOfTimeline) {
+    var tl = startOfTimeline ? this._start : this._end;
+
+    var delta = this._eventCount - this._windowLimit;
+
+    while (delta > 0) {
+        var count = startOfTimeline ? tl.advance(delta) : tl.retreat(delta);
+        if (count <= 0) {
+            // sadness. This shouldn't be possible.
+            throw new Error(
+                "Unable to unpaginate any further, but still have " +
+                    this._eventCount + " events");
+        }
+
+        delta -= count;
+        this._eventCount -= count;
+        debuglog("TimelineWindow.unpaginate: dropped " + count +
+                 " (now " + this._eventCount + ")");
+    }
+};
+
+
+/**
+ * Get a list of the events currently in the window
+ *
+ * @return {MatrixEvent[]} the events in the window
+ */
+TimelineWindow.prototype.getEvents = function() {
+    if (!this._start) {
+        // not yet loaded
+        return [];
+    }
+
+    var result = [];
+
+    for (var timeline = this._start.timeline; timeline !== null;
+        timeline = timeline.getNeighbouringTimeline(EventTimeline.FORWARDS)) {
+
+        var events = timeline.getEvents();
+        var startIndex = 0, endIndex = events.length;
+        if (timeline === this._start.timeline) {
+            startIndex = this._start.index + timeline.getBaseIndex();
+        }
+        if (timeline === this._end.timeline) {
+            endIndex = this._end.index + timeline.getBaseIndex();
+        }
+
+        for (var i = startIndex; i < endIndex; i++) {
+            result.push(events[i]);
```

Comment:
Both these `for` loops need more comments. I can follow this, but this reeks of subtle indexing errors because we aren't doing `i=0; i<events.length; i++` when pushing these events. I need you to convince me that `startIndex` and `endIndex` are going to be sane in the context of `events`.